src/livekit/MediaDevicesContext.tsx

```diff
@@ -273,6 +317,73 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
   );
 };
 
+function useControlledOutput(): MediaDeviceHandle {
+  const { available } = useObservableEagerState(
+    useObservable(() => {
+      const outputDeviceData$ = setAvailableOutputDevices$.pipe(
+        map((devices) => {
+          const deviceForEarpiece = devices.find((d) => d.forEarpiece);
+          const deviceMapTuple: [string, DeviceLabel][] = devices.map(
+            ({ id, name, isEarpiece, isSpeaker /*,isExternalHeadset*/ }) => {
+              let deviceLabel: DeviceLabel = { type: "name", name };
+              // if (isExternalHeadset) // Do we want this?
+              if (isEarpiece) deviceLabel = { type: "earpiece" };
+              if (isSpeaker) deviceLabel = { type: "default", name };
+              return [id, deviceLabel];
+            },
+          );
+          return {
+            devicesMap: new Map<string, DeviceLabel>(deviceMapTuple),
+            deviceForEarpiece,
+          };
+        }),
+      );
+
+      return combineLatest([outputDeviceData$, iosDeviceMenu$]).pipe(
+        map(([{ devicesMap, deviceForEarpiece }, iosShowEarpiece]) => {
+          let available = devicesMap;
+          if (iosShowEarpiece && !!deviceForEarpiece) {
+            available = new Map([
+              ...devicesMap.entries(),
+              [EARPIECE_CONFIG_ID, { type: "earpiece" }],
+            ]);
+          }
+          return { available, deviceForEarpiece };
+        }),
+      );
+    }),
+  );
+  const [preferredId, setPreferredId] = useSetting(audioOutputSetting);
+  useEffect(() => {
+    setOutputDevice$.subscribe((id) => {
+      if (id) setPreferredId(id);
+    });
+  }, [setPreferredId]);
+
+  const selectedId = useSelectedId(available, preferredId);
+
+  const [asEarpiece, setAsEarpiece] = useState(false);
+
+  useEffect(() => {
+    // In earpiece mode we just sent the EARPIECE_CONFIG_ID to the native code
+    // This only happens on ios where we use the native picker.
+    // So this only is needed so that ios can know if the proximity sensor should be used or not.
+    if (selectedId) window.controls.onOutputDeviceSelect?.(selectedId);
+    setAsEarpiece(selectedId === EARPIECE_CONFIG_ID);
```

Comment:
I'm finding these comments hard to follow since they're mixing a description of what EC does with a description of what downstream applications do. Is something like this still accurate?

```suggestion
    // Let the hosting application know which output device has been selected.
    // This information is probably only of interest if the earpiece mode has been
    // selected - for example, Element X iOS listens to this to determine whether it
    // should enable the proximity sensor.
    if (selectedId) window.controls.onOutputDeviceSelect?.(selectedId);
    setAsEarpiece(selectedId === EARPIECE_CONFIG_ID);
```

Reply:
very accurate and undestandable to me!